graph/src/components/store.rs

```diff
@@ -123,8 +126,40 @@ pub trait BasicStore {
     fn find(&self, query: StoreQuery) -> Result<Vec<Entity>, ()>;
 }
 
+/// Common trait for block data store implementations.
+pub trait BlockStore {
+    /// Add a new network, but only if one with this name does not already exist in the block store
+    fn add_network_if_missing(&self, network_name: &str) -> Result<(), Error>;
+
+    /// Insert some blocks into the store (or update if they are already present).
+    fn upsert_blocks(&self, network_name: &str, blocks: &[Block<Transaction>])
+        -> Result<(), Error>;
+
+    /// Try to update the head block pointer to the block with the highest block number.
+    /// Only updates pointer if there is a block with a higher block number than the current head
+    /// block, and the `ancestor_count` most recent ancestors of that block are in the store.
+    ///
+    /// If the pointer was updated, returns `Ok(vec![])`, and fires a HeadUpdateEvent.
+    ///
+    /// If no block has a number higher than the current head block, returns `Ok(vec![])`.
+    ///
+    /// If the candidate new head block had one or more missing ancestors, returns
+    /// `Ok(missing_blocks)`, where `missing_blocks` is a nonexhaustive list of missing blocks.
+    fn attempt_head_update(
+        &self,
+        network_name: &str,
+        ancestor_count: u64,
+    ) -> Result<Vec<H256>, Error>;
+
+    /// Delete blocks with block number less than `head_block_number - ancestor_count`.
+    fn delete_old_blocks(&self, network_name: &str, ancestor_count: u64) -> Result<(), Error>;
+
+    // TODO will receive from postgres "LISTEN" command
```

Comment:
Same with this commented out code, I'd also leave that out.

Reply:
Done!